File: Subsets2.py
"""
Given a collection of integers that might contain duplicates, nums, return all possible subsets.

Note:
Elements in a subset must be in non-descending order.
The solution set must not contain duplicate subsets.
For example,
If nums = [1,2,2], a solution is:

[
  [2],
  [1],
  [1,2,2],
  [2,2],
  [1,2],
  []
]
"""


class Solution:
    # @param {integer[]} nums
    # @return {integer[][]}
    def subsetsWithDup(self, nums):
        res = [[]]
        nums.sort()

        def __subsetsWithDup(idx, cur):
            for i in xrange(idx, len(nums)):
                if i > idx and nums[i] == nums[i-1]:
                    continue
                cur.append(nums[i])
                res.append(list(cur))
                __subsetsWithDup(i+1, cur)
                cur.pop()
        __subsetsWithDup(0, [])
        return res


    def subsetsWithDup2(self, nums):
        nums.sort()
        res = [[]]
        start = 0
        for i in xrange(len(nums)):
            size = len(res)
            for subset in res[start:size]:
                temp = list(subset)
                temp.append(nums[i])
                res.append(temp)
            if i < len(nums) - 1 and nums[i] == nums[i + 1]:
                start = size
            else:
                start = 0
        return res


if __name__ == '__main__':
    print Solution().subsetsWithDup([1, 2, 2, 2])

File: Leetcode-Java/Subsets2.java
import java.lang.reflect.Array;
import java.util.*;

/**
 * Created by yhf on 15/5/18.
 */

/*
Given a collection of integers that might contain duplicates, nums, return all possible subsets.

Note:
Elements in a subset must be in non-descending order.
The solution set must not contain duplicate subsets.
For example,
If nums = [1,2,2], a solution is:

[
  [2],
  [1],
  [1,2,2],
  [2,2],
  [1,2],
  []
]
*/

public class Subsets2 {
    public static List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<Integer> cur = new ArrayList<>();
        List<List<Integer>> res = new ArrayList<>();
        Set<List<Integer>> resSet = new HashSet<>();
        dfs(nums, 0, cur, res, resSet);
        return res;
    }

    public static void dfs(int[] nums, int index, List<Integer> cur, List<List<Integer>> res, Set<List<Integer>> resSet) {
        if (index == nums.length) {
            List l = new ArrayList<>(cur);
            if (!resSet.contains(l)) {
                resSet.add(l);
                res.add(l);
            }
            return;
        }
        cur.add(nums[index++]);
        dfs(nums, index, cur, res, resSet);
        cur.remove(cur.size() - 1);
        dfs(nums, index, cur, res, resSet);
    }

    public static void main(String[] args) {
        List<List<Integer>> res = subsetsWithDup(new int[]{1, 2, 2});
    }
}
